{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6cc7aa67",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import os\n",
    "from glob import glob\n",
    "import re \n",
    "\n",
    "# Import all control genotypes\n",
    "folder_path = \"/Users/rachel/Desktop/GitRepos/MachineLearning/RA_Risk/Controls\"\n",
    "file_list = glob(os.path.join(folder_path, \"*.xls\"))\n",
    "all_dfs = []\n",
    "for path in file_list:\n",
    "    df = pd.read_excel(path)\n",
    "    filename = os.path.basename(path)\n",
    "    # print(filename)\n",
    "    match = re.search(r'chr(\\d+\\d)', filename)\n",
    "    if match:\n",
    "        chromosome = match.group(1)\n",
    "    else:\n",
    "        chromosome = \"X\"\n",
    "    df['chromosome'] = chromosome\n",
    "    all_dfs.append(df)\n",
    "\n",
    "CTRL_df = pd.concat(all_dfs, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4e3cd55b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Control preprocessing\n",
    "# CTRL_df[\"condition\"] = \"CTRL\"\n",
    "# CTRL_df['MAF'] = CTRL_df.apply(lambda row: min(row['allele_A_freq'], row['allele_B_freq']), axis=1)\n",
    "CTRL_df = CTRL_df.rename(columns={'allele_A':'allele_A_CTRL', 'allele_B':'allele_B_CTRL'})\n",
    "CTRL_df.head()\n",
    "\n",
    "# We need allele counts! \n",
    "CTRL_df['allele_A_CTRL_Count'] = (CTRL_df['genotype_AA_count'] * 2) + (CTRL_df['genotype_AB_count'])\n",
    "CTRL_df['allele_B_CTRL_Count'] = (CTRL_df['genotype_BB_count'] * 2) + (CTRL_df['genotype_AB_count'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43eecc65",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2219ab18",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['dbSNP_rsID', 'allele_A', 'allele_B', 'genotype_AA_count',\n",
      "       'genotype_AB_count', 'genotype_BB_count', 'genotype_XX_count',\n",
      "       'genotype_AA_freq', 'genotype_AB_freq', 'genotype_BB_freq',\n",
      "       'allele_A_freq', 'allele_B_freq', 'allele_of_MAF', 'HWE_p_value',\n",
      "       'genotype_suspicious', 'number_of_mapped', 'chromosome',\n",
      "       'position_in_chromosome', 'duplicate_SNP', 'perlegen_rsID',\n",
      "       'perlegen_allele_1', 'perlegen_allele_2', 'perlegen_chromosome',\n",
      "       'perlegen_accession_id', 'perlegen_contig_position', 'perlegen_strand',\n",
      "       'perlegen_Assayed_sequence'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# Import RA genotypes\n",
    "RA_df = pd.read_csv(\"/Users/rachel/Desktop/GitRepos/MachineLearning/RA_Risk/Disease_Genotypes_RA.txt\", sep=\"\\t\", dtype={'chromosome':str,}, na_values=[\".-000\"])\n",
    "print(RA_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "06aa1b84",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2x/jkbf1_7146v0xqx_m99sdm000000gn/T/ipykernel_5682/2555743811.py:7: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  RA_df['genotype_suspicious'] = RA_df['genotype_suspicious'].replace('*', 1).fillna(0).astype(int)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['dbSNP_rsID', 'allele_A_RA', 'allele_B_RA', 'genotype_AA_count',\n",
       "       'genotype_AB_count', 'genotype_BB_count', 'genotype_XX_count',\n",
       "       'genotype_AA_freq', 'genotype_AB_freq', 'genotype_BB_freq',\n",
       "       'allele_A_freq', 'allele_B_freq', 'allele_of_MAF', 'HWE_p_value',\n",
       "       'genotype_suspicious', 'number_of_mapped', 'chromosome',\n",
       "       'position_in_chromosome', 'duplicate_SNP', 'perlegen_rsID',\n",
       "       'perlegen_allele_1', 'perlegen_allele_2', 'perlegen_chromosome',\n",
       "       'perlegen_accession_id', 'perlegen_contig_position', 'perlegen_strand',\n",
       "       'perlegen_Assayed_sequence', 'allele_A_comp_RA', 'allele_B_comp_RA',\n",
       "       'allele_A_RA_Count', 'allele_B_RA_Count'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def complement(nucleotide):\n",
    "    complement_dict = str.maketrans(\"ACTG\", \"TGAC\")\n",
    "    complement = nucleotide.translate(complement_dict)\n",
    "    return(complement)\n",
    "\n",
    "# RA preprocessing\n",
    "RA_df['genotype_suspicious'] = RA_df['genotype_suspicious'].replace('*', 1).fillna(0).astype(int)\n",
    "# RA_df['MAF'] = RA_df.apply(lambda row: min(row['allele_A_freq'], row['allele_B_freq']), axis=1)\n",
    "# RA_df['condition'] = \"RA\"\n",
    "\n",
    "# Strand mismatch between RA and CTRL samples found\n",
    "# Computing complement to check for strand orientation mismatch\n",
    "RA_df['allele_A_comp_RA'] = RA_df['allele_A'].apply(complement)\n",
    "RA_df['allele_B_comp_RA'] = RA_df['allele_B'].apply(complement)\n",
    "RA_df = RA_df.rename(columns={'allele_A':'allele_A_RA', 'allele_B':'allele_B_RA'})\n",
    "\n",
    "# We need allele counts! \n",
    "RA_df['allele_A_RA_Count'] = (RA_df['genotype_AA_count'] * 2) + (RA_df['genotype_AB_count'])\n",
    "RA_df['allele_B_RA_Count'] = (RA_df['genotype_BB_count'] * 2) + (RA_df['genotype_AB_count'])\n",
    "\n",
    "RA_df.head()\n",
    "RA_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0e4e9dce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbSNP_rsID</th>\n",
       "      <th>allele_A_RA</th>\n",
       "      <th>allele_B_RA</th>\n",
       "      <th>allele_A_comp_RA</th>\n",
       "      <th>allele_B_comp_RA</th>\n",
       "      <th>allele_A_CTRL</th>\n",
       "      <th>allele_B_CTRL</th>\n",
       "      <th>allele_A_RA_Count</th>\n",
       "      <th>allele_B_RA_Count</th>\n",
       "      <th>allele_A_CTRL_Count</th>\n",
       "      <th>allele_B_CTRL_Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>970</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>309</td>\n",
       "      <td>79</td>\n",
       "      <td>1441</td>\n",
       "      <td>427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2060</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>96</td>\n",
       "      <td>294</td>\n",
       "      <td>1411</td>\n",
       "      <td>457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2952</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>345</td>\n",
       "      <td>45</td>\n",
       "      <td>1614</td>\n",
       "      <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3431</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>285</td>\n",
       "      <td>103</td>\n",
       "      <td>604</td>\n",
       "      <td>1264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3444</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>352</td>\n",
       "      <td>38</td>\n",
       "      <td>131</td>\n",
       "      <td>1737</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   dbSNP_rsID allele_A_RA allele_B_RA allele_A_comp_RA allele_B_comp_RA  \\\n",
       "0         970           A           G                T                C   \n",
       "1        2060           G           A                C                T   \n",
       "2        2952           T           C                A                G   \n",
       "3        3431           C           T                G                A   \n",
       "4        3444           C           T                G                A   \n",
       "\n",
       "  allele_A_CTRL allele_B_CTRL  allele_A_RA_Count  allele_B_RA_Count  \\\n",
       "0             T             C                309                 79   \n",
       "1             A             G                 96                294   \n",
       "2             T             C                345                 45   \n",
       "3             A             G                285                103   \n",
       "4             A             G                352                 38   \n",
       "\n",
       "   allele_A_CTRL_Count  allele_B_CTRL_Count  \n",
       "0                 1441                  427  \n",
       "1                 1411                  457  \n",
       "2                 1614                  254  \n",
       "3                  604                 1264  \n",
       "4                  131                 1737  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RA_subset = RA_df[['dbSNP_rsID','allele_A_RA', 'allele_B_RA', 'allele_A_comp_RA','allele_B_comp_RA','allele_A_RA_Count','allele_B_RA_Count']]\n",
    "CTRL_subset = CTRL_df[['dbSNP_rsID','allele_A_CTRL','allele_B_CTRL','allele_A_CTRL_Count','allele_B_CTRL_Count']]\n",
    "df_clean = pd.merge(RA_subset, CTRL_subset, how=\"inner\", on=\"dbSNP_rsID\")\n",
    "df_clean.head()\n",
    "\n",
    "front_cols = ['dbSNP_rsID', 'allele_A_RA', 'allele_B_RA', 'allele_A_comp_RA', 'allele_B_comp_RA', 'allele_A_CTRL', 'allele_B_CTRL']\n",
    "other_cols = [col for col in df_clean.columns if col not in front_cols]\n",
    "organized_df = df_clean[front_cols + other_cols]\n",
    "organized_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b4342b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_mistmatch(row):\n",
    "    disease_A = row['allele_A_RA']\n",
    "    disease_B = row['allele_B_RA']\n",
    "    disease_A_rev = row['allele_A_comp_RA']\n",
    "    disease_B_rev = row['allele_B_comp_RA']\n",
    "    ctrl_A = row['allele_A_CTRL']\n",
    "    ctrl_B = row['allele_B_CTRL']\n",
    "    \n",
    "    # Case 1: alleles already match\n",
    "    if disease_A == ctrl_A and disease_B == ctrl_B:\n",
    "        return pd.Series([\"direct\"])\n",
    "    \n",
    "    # Case 2: alleles are flipped\n",
    "    elif disease_A == ctrl_B and disease_B == ctrl_A:\n",
    "        return pd.Series([\"flipped\"])\n",
    "    \n",
    "    # Case 3: Strand mismatch, direct\n",
    "    elif disease_A_rev == ctrl_A and disease_B_rev == ctrl_B:\n",
    "        return pd.Series([\"rev comp direct\"])\n",
    "    \n",
    "    # Case 4: Strand mismatch, flipped\n",
    "    elif disease_A_rev == ctrl_B and disease_B_rev == ctrl_A:\n",
    "        return pd.Series([\"rev comp flipped\"])\n",
    "    \n",
    "    # Case 5: no match (could be multiallelic, strand error, or data error)\n",
    "    else:\n",
    "        return pd.Series([\"no match\"])\n",
    "    \n",
    "\n",
    "df_clean[['match_type']] = (df_clean.apply(detect_mistmatch, axis=1))\n",
    "#df_aligned = df_clean[df_clean['match_type']!= \"no match\"]\n",
    "count = 0\n",
    "for index, row in df_clean.iterrows(): \n",
    "    if row['match_type'] == \"no match\":\n",
    "        count += 1\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "# print(count) # no discordant SNPs - all are matched as direct, flipped, or complementary alleles \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1a2afcce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbSNP_rsID</th>\n",
       "      <th>allele_A_RA</th>\n",
       "      <th>allele_B_RA</th>\n",
       "      <th>allele_A_comp_RA</th>\n",
       "      <th>allele_B_comp_RA</th>\n",
       "      <th>allele_A_RA_Count</th>\n",
       "      <th>allele_B_RA_Count</th>\n",
       "      <th>allele_A_CTRL</th>\n",
       "      <th>allele_B_CTRL</th>\n",
       "      <th>allele_A_CTRL_Count</th>\n",
       "      <th>allele_B_CTRL_Count</th>\n",
       "      <th>match_type</th>\n",
       "      <th>harmonized</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>970</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>309</td>\n",
       "      <td>79</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1441</td>\n",
       "      <td>427</td>\n",
       "      <td>rev comp direct</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2060</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>96</td>\n",
       "      <td>294</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>457</td>\n",
       "      <td>1411</td>\n",
       "      <td>flipped</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2952</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>345</td>\n",
       "      <td>45</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1614</td>\n",
       "      <td>254</td>\n",
       "      <td>direct</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3431</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>285</td>\n",
       "      <td>103</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1264</td>\n",
       "      <td>604</td>\n",
       "      <td>rev comp flipped</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3444</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>352</td>\n",
       "      <td>38</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1737</td>\n",
       "      <td>131</td>\n",
       "      <td>rev comp flipped</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   dbSNP_rsID allele_A_RA allele_B_RA allele_A_comp_RA allele_B_comp_RA  \\\n",
       "0         970           A           G                T                C   \n",
       "1        2060           G           A                C                T   \n",
       "2        2952           T           C                A                G   \n",
       "3        3431           C           T                G                A   \n",
       "4        3444           C           T                G                A   \n",
       "\n",
       "   allele_A_RA_Count  allele_B_RA_Count allele_A_CTRL allele_B_CTRL  \\\n",
       "0                309                 79             T             C   \n",
       "1                 96                294             G             A   \n",
       "2                345                 45             T             C   \n",
       "3                285                103             A             G   \n",
       "4                352                 38             A             G   \n",
       "\n",
       "   allele_A_CTRL_Count  allele_B_CTRL_Count        match_type harmonized  \n",
       "0                 1441                  427   rev comp direct        Yes  \n",
       "1                  457                 1411           flipped        Yes  \n",
       "2                 1614                  254            direct        Yes  \n",
       "3                 1264                  604  rev comp flipped        Yes  \n",
       "4                 1737                  131  rev comp flipped        Yes  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clean.head()\n",
    "\n",
    "def harmonize_alleles(df):\n",
    "    for index, row in df.iterrows():\n",
    "        match = row['match_type']\n",
    "        if match == \"direct\":\n",
    "            pass\n",
    "        if match == \"flipped\":\n",
    "            df.at[index, 'allele_A_CTRL'], df.at[index, 'allele_B_CTRL'] = row['allele_B_CTRL'], row['allele_A_CTRL']\n",
    "            df.at[index, 'allele_A_CTRL_Count'], df.at[index, 'allele_B_CTRL_Count'] = row['allele_B_CTRL_Count'], row['allele_A_CTRL_Count']\n",
    "        elif match == \"rev comp direct\":\n",
    "            df.at[index, 'allele_A_CTRL'], df.at[index, \"allele_B_CTRL\"] = row['allele_A_comp_RA'], row['allele_B_comp_RA']\n",
    "        elif match == \"rev comp flipped\":\n",
    "            df.at[index, 'allele_A_CTRL'], df.at[index, \"allele_B_CTRL\"] = row['allele_B_comp_RA'], row['allele_A_comp_RA']\n",
    "            df.at[index, 'allele_A_CTRL_Count'], df.at[index, 'allele_B_CTRL_Count'] = row['allele_B_CTRL_Count'], row['allele_A_CTRL_Count']\n",
    "        else:\n",
    "            f\"Match type not recognized: {index}, {row}, {match}\"\n",
    "    df['harmonized'] = \"Yes\"\n",
    "    return df\n",
    "\n",
    "df_harmonized = harmonize_alleles(df_clean)\n",
    "df_harmonized.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "90e1a2d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import fisher_exact\n",
    "import statsmodels\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "def fisher_test(df):\n",
    "    results = []\n",
    "    for index, row in df.iterrows():\n",
    "        table = [row['allele_A_RA_Count'], row['allele_B_RA_Count']], [row['allele_A_CTRL_Count'], row['allele_B_CTRL_Count']]\n",
    "        oddr, pval = fisher_exact(table)\n",
    "        results.append({\n",
    "            'SNP_rsID': row['dbSNP_rsID'],\n",
    "            'odds_ratio': oddr,\n",
    "            'p_value':pval\n",
    "        })\n",
    "    fisher_df = pd.DataFrame(results)\n",
    "    return fisher_df\n",
    "        \n",
    "fisher_df = fisher_test(df_harmonized)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cadb65aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37301\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dbSNP_rsID             11227065\n",
       "allele_A_RA                   C\n",
       "allele_B_RA                   A\n",
       "allele_A_comp_RA              G\n",
       "allele_B_comp_RA              T\n",
       "allele_A_RA_Count           376\n",
       "allele_B_RA_Count            10\n",
       "allele_A_CTRL                 C\n",
       "allele_B_CTRL                 A\n",
       "allele_A_CTRL_Count        1857\n",
       "allele_B_CTRL_Count          11\n",
       "match_type              flipped\n",
       "harmonized                  Yes\n",
       "Name: 37301, dtype: object"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fisher_df.head()\n",
    "padj = []\n",
    "for index, row in fisher_df.iterrows():\n",
    "    p_adj = multipletests(row['p_value'], method='fdr_bh')[1]\n",
    "    padj.append(p_adj)\n",
    "fisher_df[\"p_adj\"] = padj\n",
    "fisher_df.head()\n",
    "# Annotations for snps with p_adj of 0 \n",
    "# df_filtered_00['SNP_rsID'] = 'rs' + df_filtered_00['SNP_rsID'].astype(str)\n",
    "# df_00_sorted = df_filtered_00.sort_values(by='odds_ratio', ascending=False)\n",
    "# df_00_sorted.head()\n",
    "\n",
    "\n",
    "idx = df_harmonized[df_harmonized['dbSNP_rsID'] == 11227065].index[0]\n",
    "print(idx)\n",
    "df_harmonized.iloc[idx]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c434e6c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP_rsID</th>\n",
       "      <th>chi2</th>\n",
       "      <th>p_value</th>\n",
       "      <th>deg_freedom</th>\n",
       "      <th>expected</th>\n",
       "      <th>p_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>970</td>\n",
       "      <td>1.013008</td>\n",
       "      <td>0.314183</td>\n",
       "      <td>1</td>\n",
       "      <td>[[300.9751773049645, 87.02482269503547], [1449...</td>\n",
       "      <td>[0.31418340906441816]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2060</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>[[95.51372896368467, 294.4862710363153], [457....</td>\n",
       "      <td>[1.0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2952</td>\n",
       "      <td>1.018105</td>\n",
       "      <td>0.312969</td>\n",
       "      <td>1</td>\n",
       "      <td>[[338.35695305580157, 51.6430469441984], [1620...</td>\n",
       "      <td>[0.3129690377964257]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3431</td>\n",
       "      <td>4.735934</td>\n",
       "      <td>0.029539</td>\n",
       "      <td>1</td>\n",
       "      <td>[[266.40602836879435, 121.59397163120568], [12...</td>\n",
       "      <td>[0.029538800542831384]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3444</td>\n",
       "      <td>3.091395</td>\n",
       "      <td>0.078707</td>\n",
       "      <td>1</td>\n",
       "      <td>[[360.81045172719223, 29.189548272807794], [17...</td>\n",
       "      <td>[0.07870732241078082]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   SNP_rsID      chi2   p_value  deg_freedom  \\\n",
       "0       970  1.013008  0.314183            1   \n",
       "1      2060  0.000000  1.000000            1   \n",
       "2      2952  1.018105  0.312969            1   \n",
       "3      3431  4.735934  0.029539            1   \n",
       "4      3444  3.091395  0.078707            1   \n",
       "\n",
       "                                            expected                   p_adj  \n",
       "0  [[300.9751773049645, 87.02482269503547], [1449...   [0.31418340906441816]  \n",
       "1  [[95.51372896368467, 294.4862710363153], [457....                   [1.0]  \n",
       "2  [[338.35695305580157, 51.6430469441984], [1620...    [0.3129690377964257]  \n",
       "3  [[266.40602836879435, 121.59397163120568], [12...  [0.029538800542831384]  \n",
       "4  [[360.81045172719223, 29.189548272807794], [17...   [0.07870732241078082]  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.stats import chi2_contingency\n",
    "\n",
    "def chi2_test(df):\n",
    "    results = []\n",
    "    for index, row in df.iterrows():\n",
    "        table = [row['allele_A_RA_Count'], row['allele_B_RA_Count']],[row['allele_A_CTRL_Count'], row['allele_B_CTRL_Count']]\n",
    "        chi2, p, dof, expected = chi2_contingency(table)\n",
    "        results.append({\n",
    "            'SNP_rsID': row['dbSNP_rsID'],\n",
    "            'chi2': chi2,\n",
    "            'p_value':p,\n",
    "            'deg_freedom':dof,\n",
    "            'expected':expected\n",
    "        })\n",
    "    chi2_df = pd.DataFrame(results)\n",
    "    return chi2_df\n",
    "\n",
    "chi2_df = chi2_test(df_harmonized)\n",
    "\n",
    "padj = []\n",
    "for index, row in chi2_df.iterrows():\n",
    "    p_adj = multipletests(row['p_value'], method='fdr_bh')[1]\n",
    "    padj.append(p_adj)\n",
    "chi2_df[\"p_adj\"] = padj\n",
    "chi2_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "86e10965",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered = chi2_df[chi2_df['p_adj']<0.05]\n",
    "len(filtered)\n",
    "filtered_test_df = filtered.head(n=40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2c27dff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Converting SNP ids to rs IDs for NCBI search\n",
    "snp_list = filtered_test_df['SNP_rsID'].astype(str).tolist()\n",
    "rs_ids = []\n",
    "for snp in snp_list:\n",
    "    rs_id = \"rs\" + snp\n",
    "    rs_ids.append(rs_id)\n",
    "\n",
    "\n",
    "# API Calls to retrieve data for each SNP\n",
    "import requests\n",
    "import re\n",
    "import time\n",
    "\n",
    "# 1. Create batches of 100 \n",
    "def create_batch(iterable, n=20):\n",
    "    for i in range(0, len(iterable), n):\n",
    "        yield iterable[i:i+n]\n",
    "\n",
    "# Convert rsID to UID\n",
    "def rs_to_uid(snp_rsID_list):\n",
    "    db = \"snp\"\n",
    "    base = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/\"\n",
    "    term = \" OR \".join(snp_rsID_list)\n",
    "    parameters = {\n",
    "        \"db\":db,\n",
    "        \"term\": term,\n",
    "        \"retmode\":\"xml\"\n",
    "    }\n",
    "    url = f\"{base}esearch.fcgi\"\n",
    "    response = requests.get(url, parameters)\n",
    "    response.raise_for_status()\n",
    "    with open(\"snp_uid.xml\", \"a\") as file:\n",
    "        file.write(response.text)\n",
    "  \n",
    "for chunk in create_batch(rs_ids, n=20):\n",
    "    uids = rs_to_uid(chunk)\n",
    "    time.sleep(1)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "e5c979fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['386605925', '386575067', '386431412', '111190448', '61074242', '60993869', '60931066', '60802411', '60666408', '60501608', '60352433', '60331063', '60030157', '59783138', '58976013', '58870618', '58655838', '58339502', '58159257', '57834999', '2100230641', '2100230631', '2100230624', '2100230612', '386624444', '386623847', '386623308', '386606604', '386606419', '61511205', '61506800', '60996781', '60551916', '60244893', '60071076', '59216045', '58844537', '58697252', '58667553', '58617118']\n"
     ]
    }
   ],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import lxml\n",
    "import re\n",
    "uid_query_list = []\n",
    "with open(\"snp_uid.xml\", \"r\") as file: \n",
    "    contents = file.read()\n",
    "    clean_contents = re.sub(r'<\\?xml.*?\\?>', \"\", contents)\n",
    "    clean_contents = re.sub(r'<!DOCTYPE.*?>', \"\", clean_contents)\n",
    "    wrapped_contents = f\"<root>{clean_contents}</root>\"\n",
    "    soup = BeautifulSoup(wrapped_contents, features=\"xml\")\n",
    "    id_list = soup.find_all(\"Id\")\n",
    "    for id in id_list: \n",
    "        uid_query_list.append(id.text)\n",
    "\n",
    "print(uid_query_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "65e80ba2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\\n<!DOCTYPE ePostResult PUBLIC \"-//NLM//DTD epost 20090526//EN\" \"https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20090526/epost.dtd\"><ePostResult>\\n\\t<ERROR>Empty ID list; Nothing to store</ERROR>\\n</ePostResult>\\n'"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def snpdb_call(uid_list):\n",
    "    \"\"\" \n",
    "    API call to retrieve data for all SNPs in the given list.\n",
    "    :param snp_rsID_list: list of all SNPs to query. SNPs must be preceeded by rs (example: rs3431)\n",
    "    \"\"\"\n",
    "    db = \"snp\"\n",
    "    base = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/\"\n",
    "\n",
    "    # POST request\n",
    "    post_data = {\n",
    "        \"db\": db,\n",
    "        \"ids\":uid_list\n",
    "    }\n",
    "\n",
    "    epost_url = f\"{base}epost.fcgi\"\n",
    "    epost = requests.post(epost_url, post_data)\n",
    "    return epost.text\n",
    "\n",
    "    # if epost.status_code != 200:\n",
    "    #     raise Exception(f\"Epost request failed. Status code: {epost.status_code}\")\n",
    "    \n",
    "    # web_match = re.search(r\"<WebEnv>(\\S+)</WebEnv>\", str(epost))\n",
    "    # web = web_match.group(1) if web_match else None\n",
    "    # if web == None:\n",
    "    #     raise Exception(\"Web value not found.\")\n",
    "    \n",
    "    # key_match = re.search(r\"<QueryKey>(\\d+)</QueryKey>\", str(epost))\n",
    "    # key = key_match.group(1) if key_match else None\n",
    "    # if key == None: \n",
    "    #     raise Exception(\"Key value not found.\")    \n",
    "    \n",
    "    # efetch_url = f\"{base}efetch.fcgi?db={db}&query_key={key}&WebEnv={web}&retmode=xml\"\n",
    "    # efetch = requests.get(efetch_url)\n",
    "\n",
    "    # if efetch.status_code != 200:\n",
    "    #     raise Exception(f\"Efetch request failed. Status code: {efetch.status_code}\")\n",
    "    \n",
    "    # return efetch.text\n",
    "\n",
    "snpdb_call(uid_query_list)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
